### PR TITLE
ENH New steps to creat groups and members with permissions

### PR DESCRIPTION
**Pull request by sabina-talipova**
### Description
- New methods to generate Group `generateGroupWithPermission()` and Members `generateMemeber()` separately
- Updated  `generateMemberWithPermission()`
- New step `iAmLoggedInAsMemeberOfGroup()` - `Given I am logged in as a memeber of "ADMIN" group`

### Parent issue

- https://github.com/silverstripe/silverstripe-behat-extension/issues/220

### CODE REVIEW

File: src/Context/LoginContext.php [hunk 1]
```diff
@@ -53,6 +54,26 @@ public function iAmLoggedInWithPermissions($permCode)
...
+    /**
+     * Creates and login as a member of a group with the correct permissions.
+     * Example: Given I am logged in as a memeber of "ADMIN" group
+     *
+     * @Given /^I am logged in as a memeber of "([^"]*)" group$/
+     * @param string $permCode
+     */
+    public function iAmLoggedInAsMemeberOfGroup($permCode)
+    {
+        $group = Group::get()->filter('Title', "$permCode group")->first();
```
Comment (GuySartorelli) [inline src/Context/LoginContext.php]:
```suggestion
    public function iAmLoggedInAsMemeberOfGroup($groupName)
    {
        $group = Group::get()->filter('Title', "$groupName")->first();
```
Remove the necessity to have "group" at the end of the group name.

File: src/Context/LoginContext.php [hunk 2]
```diff
@@ -53,6 +54,26 @@ public function iAmLoggedInWithPermissions($permCode)
...
+        if (!$group) {
+            throw new LogicException("Group $group does not exist");
+        }
+
+        $email = "{$permCode}@example.org";
+        $password = 'Secret!123';
+        $this->generateMemeber($email, $password, $group, $permCode);
```
Comment (GuySartorelli) [inline src/Context/LoginContext.php]:
```suggestion
        $this->generateMember($email, $password, $group, $permCode);
```

Comment (GuySartorelli) [inline src/Context/LoginContext.php]:
```suggestion
            throw new LogicException("Group $groupName does not exist");
        }

        $email = "{$groupName}@example.org";
        $password = 'Secret!123';
        $this->generateMemeber($email, $password, $group, $groupName);
```

File: src/Context/LoginContext.php [hunk 3]
```diff
@@ -245,22 +266,48 @@ public function stepPasswordForEmailShouldBe($id, $password)
      * @return Member
      */
     protected function generateMemberWithPermission($email, $password, $permCode)
+    {
+        $group = $this->generateGroupWithPermission($permCode);
+
+        return $this->generateMemeber($email, $password, $group, $permCode);
```
Comment (GuySartorelli) [inline src/Context/LoginContext.php]:
```suggestion
        return $this->generateMember($email, $password, $group, $permCode);
```

File: src/Context/LoginContext.php [hunk 4]
```diff
@@ -245,22 +266,48 @@ public function stepPasswordForEmailShouldBe($id, $password)
...
+    /**
+     * Get or generate a member with the given permission code and permission group
+     *
+     * @param string $email
+     * @param string $password
+     * @param object $group
+     * @param string $permCode
+     * @return Member
+     */
+    protected function generateMemeber($email, $password, $group, $permCode)
```
Comment (GuySartorelli) [inline src/Context/LoginContext.php]:
```suggestion
    protected function generateMemeber($email, $password, $group, $identifier)
```
It's not necessarily a permission code.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Context/LoginContext.php [hunk 5]
```diff
@@ -245,22 +266,48 @@ public function stepPasswordForEmailShouldBe($id, $password)
...
+    {
+        $group = $this->generateGroupWithPermission($permCode);
+
+        return $this->generateMember($email, $password, $group, $permCode);
+    }
+
+    /**
+     * Get or generate a group with the given permission code
+     *
+     * @param string $groupName
```
Comment (GuySartorelli) [inline src/Context/LoginContext.php]:
```suggestion
     * @param string $permCode
```
This is still the permission code.

File: src/Context/LoginContext.php [hunk 6]
```diff
@@ -245,22 +266,48 @@ public function stepPasswordForEmailShouldBe($id, $password)
...
+     * @return Member
+     */
+    protected function generateGroupWithPermission($permCode)
     {
         // Get or create group
-        $group = Group::get()->filter('Title', "$permCode group")->first();
+        $group = Group::get()->filter('Title', "$permCode")->first();
         if (!$group) {
             $group = Group::create();
+            $group->Title = "$permCode";
```
Comment (GuySartorelli) [inline src/Context/LoginContext.php]:
```suggestion
        $group = Group::get()->filter('Title', "$permCode group")->first();
        if (!$group) {
            $group = Group::create();
            $group->Title = "$permCode group";
```
We don't want to change the behaviour of generateMemberWithPermission() - that can still add "group" to the end of the group name.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Context/LoginContext.php [hunk 7]
```diff
@@ -53,6 +54,26 @@ public function iAmLoggedInWithPermissions($permCode)
         $this->stepILogInWith($email, $password);
     }
 
+    /**
+     * Creates and login as a member of a group with the correct permissions.
+     * Example: Given I am logged in as a memeber of "ADMIN" group
+     *
+     * @Given /^I am logged in as a memeber of "([^"]*)" group$/
```
Comment (GuySartorelli) [inline src/Context/LoginContext.php]:
```suggestion
     * Example: Given I am logged in as a member of "ADMIN" group
     *
     * @Given /^I am logged in as a member of "([^"]*)" group$/
```
Sorry, I should have noticed this typo last time.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

**Review by GuySartorelli — APPROVED** [review]
Looks good - hold off merging until at least one of the related PRs is ready.

File: src/Context/LoginContext.php [hunk 8]
```diff
@@ -53,6 +54,26 @@ public function iAmLoggedInWithPermissions($permCode)
...
     }
 
+    /**
+     * Creates and login as a member of a group with the correct permissions.
+     * Example: Given I am logged in as a memeber of "ADMIN" group
+     *
+     * @Given /^I am logged in as a member of "([^"]*)" group$/
+     * @param string $groupName
+     */
+    public function iAmLoggedInAsMemeberOfGroup($groupName)
```
Comment (GuySartorelli) [inline src/Context/LoginContext.php]:
```suggestion
     * Example: Given I am logged in as a member of "ADMIN" group
     *
     * @Given /^I am logged in as a member of "([^"]*)" group$/
     * @param string $groupName
     */
    public function iAmLoggedInAsMemberOfGroup($groupName)
```

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Context/LoginContext.php (hunk 8 above)
Comment (sabina-talipova) [reply src/Context/LoginContext.php]: DONE

**Review by GuySartorelli — APPROVED** [review]

**Pull request merged**